contrib_bots/lib/github_detail.py

```diff
@@ -0,0 +1,134 @@
+
+    def __init__(self):
+        config = six.moves.configparser.ConfigParser()
+        config.readfp(open(self.config_path))
+        if config.get('Github', 'repo'):
+            self.repo = config.get('Github', 'repo')
+            if config.get('Github', 'owner'):
+                self.owner = config.get('Github', 'owner')
+            else:
+                # Allowing undefined default repos would require multiple triage_message regexs.
+                # It's simpler to require them to be defined.
+                sys.exit('Default owner not defined')
+        else:
+            sys.exit('Default repo not defined')
+
+    def usage(self):
+        # type: () -> None
+        return ("This plugin displays details on github issues and pull requests. "
+                "To refrence an issue or pr type @github-detail then anytime in the message it's id like '#2561'."
+                "If you want to refrence an issues from a repository other than the default use"
+                "the user/repo#id form.")
+
+    def triage_message(self, message, client):
+        # type: () -> bool
+        # Check the message contains a username mention and #<number>
+        # replying to another bot
+        if (re.search("(?:@(?:\*\*)" + re.escape(client.full_name) + ").+?#\d+", message['content']) and
+                not message['sender_email'].endswith('-bot@zulip.com')):
+            return True
+        return False
+
+    def detail_issue(self, issue):
+        # type: (Dict[Text, Union[Text, int, bool]]) -> Text
+
+        id = issue['number']
+        title = issue['title']
+        link = issue['html_url']
+        # Truncate if longer than 200 characters.
+        description = (issue['body'][:200] +
+                       '..') if len(issue['body']) > 202 else issue['body']
+        status = issue['state'].title()
+
+        return '**[{id} | {title}]({link})** - **{status}**\n```quote\n{description}\n```'\
+            .format(id=id, title=title, link=link, status=status, description=description)
+
+    def detail_pull(self, pull):
+        # type: (Dict[Text, Union[Text, int, bool, Dict) -> Text
+
+        id = pull['number']
+        title = pull['title']
+        link = pull['html_url']
+        # Truncate if longer than 200 characters.
+        description = (pull['body'][:200] +
+                       '..') if len(pull['body']) > 202 else pull['body']
+        status = pull['state'].title()
+
+        message = """**[{id} | {title}]({link})** - **{status}**\n```quote\n{description}\n```""".format(
+            id=id, title=title, link=link, status=status, description=description)
+        return message
+
+    def get_details(self, owner, repo, id):
+        # type: (Text, Text, Text) -> Dict[Text, Union[Text, Int, Bool]]
+        # Gets the details of an issues or pull request
+
+        # Try to get an issue, try to get a pull if that fails
+        r = requests.get(
```

Comment:
What would the best behaviour be? Maybe divide into fatal and non-fatal exceptions then log and crash or log and continue?